src/EditorFeatures/TestUtilities/LanguageServer/AbstractLanguageServerProtocolTests.cs

```diff
@@ -359,23 +344,33 @@ private async Task<TestLspServer> CreateTestLspServerAsync(TestWorkspace workspa
             InitializationOptions? initializationOptions = null)
         {
             var lspOptions = initializationOptions ?? new InitializationOptions();
-            var exportProvider = Composition.ExportProviderFactory.CreateExportProvider();
-            if (lspOptions.OptionUpdater != null)
-            {
-                var globalOptions = exportProvider.GetExportedValue<IGlobalOptionService>();
-                lspOptions.OptionUpdater(globalOptions);
-            }
 
-            var workspace = TestWorkspace.Create(XElement.Parse(xmlContent), openDocuments: false, exportProvider: exportProvider, workspaceKind: workspaceKind);
-            workspace.TryApplyChanges(workspace.CurrentSolution.WithAnalyzerReferences(new[] { TestAnalyzerReferences }));
+            var workspace = CreateWorkspace(lspOptions, workspaceKind);
+
+            workspace.InitializeDocuments(XElement.Parse(xmlContent), openDocuments: false);
+            workspace.TryApplyChanges(workspace.CurrentSolution.WithAnalyzerReferences(new[] { CreateTestAnalyzersReference() }));
 
             // Important: We must wait for workspace creation operations to finish.
             // Otherwise we could have a race where workspace change events triggered by creation are changing the state
             // created by the initial test steps. This can interfere with the expected test state.
             await WaitForWorkspaceOperationsAsync(workspace);
+
             return await TestLspServer.CreateAsync(workspace, lspOptions);
         }
 
+        internal TestWorkspace CreateWorkspace(InitializationOptions? options, string? workspaceKind)
+        {
+            var workspace = new TestWorkspace(Composition, workspaceKind, configurationOptions: new WorkspaceConfigurationOptions(EnableOpeningSourceGeneratedFiles: true));
+            options?.OptionUpdater?.Invoke(workspace.GetService<IGlobalOptionService>());
+
+            workspace.GetService<LspWorkspaceRegistrationService>().Register(workspace);
+
+            var solutionCrawlerRegistrationService = (SolutionCrawlerRegistrationService)workspace.Services.GetRequiredService<ISolutionCrawlerRegistrationService>();
```

Comment:
would be good to add a quick comment why we do this - essentially so we can set global options before creating the solution crawler (which is required to initialize the diagnostics incremental analyzer).

will be glad to get rid of this once we turn on pull fully!